### ISSUE TITLE
Race condition between state machine role creation and use. Long time to propagate the role.

### ISSUE THREAD

**Issue by vishalGitAcc** [terraform-providers/terraform-provider-aws#7893]
**Terraform Version**
Terraform version 2.1.0

**Problem**: The state machine exec role takes time to propagate and an explicit “depends on” constraint does not help.

**Terraform Configuration Files**
```
data "aws_iam_policy_document" "states_exec_role_document" {
  statement {
    actions = ["sts:AssumeRole"]

    principals {
      type        = "Service"
      identifiers = ["states.amazonaws.com"]
    }
  }
}

resource "aws_iam_role" "states_exec_role" {
  name               = "${local.stack_name}-${local.pod}-statesExecRole"
  assume_role_policy = "${data.aws_iam_policy_document.states_exec_role_document.json}"
}

resource "aws_sfn_state_machine" "state_machine" {
 depends_on = ["aws_iam_role.states_exec_role"]
 name       = "${local.stack_name}-${local.pod}-state-machine"
 definition = "${data.template_file.definition.rendered}"
 role_arn   = "${aws_iam_role.states_exec_role.arn}"
}

```
**Crash Output**
"* aws_sfn_state_machine.state_machine: Error creating Step Function State Machine: AccessDeniedException: Neither the global service principal states.amazonaws.com, nor the regional one is authorized to assume the provided role.
    status code: 400, request id: b26b3cc8-442d-11e9-83bb-03aa166dd894"

**Expected Behavior**
Apply complete! Resources: X added, 0 changed, 0 destroyed.

**Actual Behavior**
Fails to apply because iam roles takes time to propagate across regions. Apply succeeds on second application. It takes a few seconds for the IAM role to be available. 

**Steps to Reproduce**
terraform apply

**Additional context**

**Following hack addresses the problem, but this is still a hack.**

```
resource "null_resource" "delay" {
 provisioner "local-exec" {
   command = "sleep 30"
 }
 triggers = {
   "states_exec_role" = "${aws_iam_role.states_exec_role.arn}"
 }
}

resource "aws_sfn_state_machine" "state_machine" {
 name       = "${local.stack_name}-${local.pod}-state-machine"
 definition = "${data.template_file.definition.rendered}"
 role_arn   = "${aws_iam_role.states_exec_role.arn}"
 depends_on = ["null_resource.delay"]
}

Looking for a better solution to address the race condition.

**Comment by sstoeckel** [comment]
I got an equal issue. The Problem is the State Machine will be created before the IAM Role and IAM Role Policy Attachment.

**Comment by mauza** [comment]
Experiencing the same issue. I've built a few step functions in the past and never ran into this...

**Comment by bundy97** [comment]
Experiencing the same issue

**Comment by prokvk** [comment]
same issue here

**Comment by Vilaggio** [comment]
Same issue here. I attempted to add a trust relationship but that error continues...

**Comment by Vilaggio** [comment]
I fixed mine by going back and adding the correct IAM role in the state machine.

![2](https://user-images.githubusercontent.com/43763535/64962465-81a6ce80-d865-11e9-853e-10def5320038.JPG)

**Comment by bflad** [comment]
The fix to allow the `aws_sfn_state_machine` resource to retry IAM Role associated failures on creation has been merged and will release with version 2.69.0 of the Terraform AWS Provider, likely tomorrow. Thanks to @DrFaust92 for the implementation. 👍
